### PR TITLE
Setup random

### PR DESCRIPTION
**Pull request by geoffthemedio**
Adds random setting to various galaxy setup options. Involves non-trivial Python script modifications that probably could use some cleaning up, but runs for me.

### CODE REVIEW

File: default/python/turn_events/turn_events.py [hunk 1]
```diff
@@ -6,6 +6,66 @@
 from universe_tables import MONSTER_FREQUENCY
 
 
+
+# tuples of galaxy setup options to pick from randomly
+low_to_high =  (fo.galaxySetupOption.low, fo.galaxySetupOption.medium, fo.galaxySetupOption.high)
+none_to_high = (fo.galaxySetupOption.low, fo.galaxySetupOption.medium, fo.galaxySetupOption.high, fo.galaxySetupOption.none)
```
Comment (Cjkjvfnby) [inline default/python/turn_events/turn_events.py]: `low_to_none`?

File: default/python/turn_events/turn_events.py [hunk 2]
```diff
@@ -6,6 +6,66 @@
...
+        self.shape = galaxy_setup_data.shape
+        self.__age = galaxy_setup_data.age
+        self.__starlane_frequency = galaxy_setup_data.starlaneFrequency
+        self.__planet_density = galaxy_setup_data.planetDensity
+        self.__specials_frequency = galaxy_setup_data.specialsFrequency
+        self.__monster_frequency = galaxy_setup_data.monsterFrequency
+        self.__native_frequency = galaxy_setup_data.nativeFrequency
+        self.max_ai_aggression = galaxy_setup_data.maxAIAggression
+    
+    def age(self):
```
Comment (Cjkjvfnby) [inline default/python/turn_events/turn_events.py]: Convert to property?

File: default/python/turn_events/turn_events.py [hunk 3]
```diff
@@ -6,6 +6,66 @@
...
+        #print("stored age: " + str(self.__age))
+        if (self.__age != fo.galaxySetupOption.random):
+            return self.__age
+        return choice(low_to_high)
+
+    def starlaneFrequency(self):
```
Comment (Cjkjvfnby) [inline default/python/turn_events/turn_events.py]: python methods should be named in `snake_case`

File: default/python/turn_events/turn_events.py [hunk 4]
```diff
@@ -6,6 +6,66 @@
 from universe_tables import MONSTER_FREQUENCY
 
 
+
+# tuples of galaxy setup options to pick from randomly
+low_to_high =  (fo.galaxySetupOption.low, fo.galaxySetupOption.medium, fo.galaxySetupOption.high)
+none_to_high = (fo.galaxySetupOption.low, fo.galaxySetupOption.medium, fo.galaxySetupOption.high, fo.galaxySetupOption.none)
+
+class PyGalaxySetupData:
```
Comment (Cjkjvfnby) [inline default/python/turn_events/turn_events.py]: Now we have two `PyGalaxySetupData` classes.  Why?

File: default/python/turn_events/turn_events.py (hunk 1 above)
Comment (geoffthemedio) [inline default/python/turn_events/turn_events.py]: Don't understand the question...

File: default/python/turn_events/turn_events.py (hunk 2 above)
Comment (geoffthemedio) [inline default/python/turn_events/turn_events.py]: example?

File: default/python/turn_events/turn_events.py (hunk 4 above)
Comment (geoffthemedio) [inline default/python/turn_events/turn_events.py]: I needed the same functionality somewhere else, but don't know where/how to move the code to be accessible from both places.

As I said, the Python needs cleaning up, preferably by someone who knows how best to do so, which is not me.

**Comment by Vezzra** [comment]
> As I said, the Python needs cleaning up, preferably by someone who knows how best to do so, which is not me.

I'll look into this, will try to get it done this week.

File: default/python/turn_events/turn_events.py (hunk 1 above)
Comment (Cjkjvfnby) [inline default/python/turn_events/turn_events.py]: It was suggestion to rename variable.
Previous variable named form first element of tuple to last, this named form last to last but one. I see inconsistency here.

File: default/python/turn_events/turn_events.py (hunk 2 above)
Comment (Cjkjvfnby) [inline default/python/turn_events/turn_events.py]:
```
@property
def age(self):
```

Usage:

```
# galaxy_age = gsd.age()
galaxy_age = gsd.age
```

```
```

File: default/python/turn_events/turn_events.py (hunk 1 above)
Comment (geoffthemedio) [inline default/python/turn_events/turn_events.py]: the order of the items is irrelevant; it just needs to pick one. the name is was chosen to convey that all, or all but none, of the possible values were included.

**Comment by Vezzra** [comment]
Ok, I've taken a brief look at this, and I don't think this implementation will work as intended. I guess this PR has been made in an attempt to implement this one request on the forum where one guy suggested to to offer a "random" option for several settings (like the galaxy shape already has), to make for a more challenging game because you don't know what to expect.

Assuming that's the case, and if I understood that proposal correctly, these "random" options should work like the "random" galaxy shape - instead of you picking the setting, the game picks one randomly for you.

In this case what we want is that one option gets randomly chosen, but after that the game should stick to that option and continue as if the player had chosen it. However, the way that is implemented in this PR the random choice is triggered _each time the setting is accessed_.

Which would be fine if each setting that offers the "random" option would be queried only once, but we can't rely on that, as the redundant implementation of the PyGalaxySetupData class and use in turn_events.py shows. Each time the setting gets accessed, it will randomly re-determine the actual option.

Moreover, the galaxy setup options are also exposed to FOCS. The FOCS interface will of course access the GalaxySetupData instance on the C++ side, which has nothing to do with anything implemented on the Python side (actually yielding the "random" option instead of a randomly picked "real" option).

I don't think that this has been the idea. I think the only way to implement that is to move the handling of the random option into the backend C++ code, the GalaxySetupData struct to be precise. This structure needs to be extended to hold the original settings the player has chosen (those are needed to be shown to the player in the pedia, and maybe in some other cases?), and the actual settings the game uses and exposes to FOCS and Python. The actual random picking can be triggered by a dedicated member function, or when the setting is first accessed (will require wrapping access to the settings with getters and setters).

@geoffthemedio, correct me if I got anything wrong here?

**Comment by geoffthemedio** [comment]
Re-picking one of the options each time was intentional. It probably doesn't actually matter for the universe generation scripts though, as other than for some logging calls, the randomized galaxy setup info doesn't appear to be accessed repeatedly. I might have missed something, though.

I didn't consider the use of the galaxy setup settings in content scripts, though that could be easily made to work the same way in Variable<int>::Eval

Part of the idea was to keep the actual choice unknown to the player, rather than picking and displaying one of the options for the player.

**Comment by geoffthemedio** [comment]
Admittedly, having a new value each times it's checked for several of the setup options that determine the probabilities of things happening in scripts will produce results somewhat different from what might be expected.

**Comment by Vezzra** [comment]
As this has turned into a design discussion, I've opened a [thread on the forums](http://www.freeorion.org/forum/viewtopic.php?f=6&t=9958). I've quoted the design discussion related comments that have been made here already in the OP of that thread.

Please, everyone, continue the discussion there, I will make further replies there too.

**Comment by geoffthemedio** [comment]
Reworked resolution of random to within C++

**Comment by Vezzra** [comment]
I tested your reworked resolution and found what I think is a major flaw/shortcoming: the way the random pick is determined by the seed hash and the number of options to pick from produces the same pick for all settings that have the same number of options to pick from.

Meaning, age, starlane frequency and planet density would _all_ be either low, medium or high, and specials, monster and native frequency would _all_ be either none, low, medium or high. That's not really sufficiently "random" in my eyes, so I came up with an alternative approach.

I've pushed that alternative implementation to a new branch ("revised-random-setup-option-picking") and created a PR for it: #485 (yeah, that's a PR on a PR :wink:). So you can take a look and review it.

Is that an acceptable approach for you?

**Comment by geoffthemedio** [comment]
All that extra data storage and transferring seems unnecessary. It would be substantially simpler to append something different to the seed before hashing in each of the getters...

**Comment by Vezzra** [comment]
Good idea, and in theory it works really nice, but when testing that on OSX _and_ Linux now, I ran into the next issue (and this one is of the I'm-pulling-out-my-hair kind): the hashes `boost::hash`produces are **_not**_ platform independent. I get different hashes for the same base string on OSX and Linux (and therefore of course completely different maps).

I added a debug log output line to the GetIdx function:

```
diff --git a/util/MultiplayerCommon.cpp b/util/MultiplayerCommon.cpp
index 44612ef..4267964 100644
--- a/util/MultiplayerCommon.cpp
+++ b/util/MultiplayerCommon.cpp
@@ -124,6 +124,7 @@ namespace {
     int GetIdx(const T1& enum_vals_count, const T2& seed) {
         static boost::hash<T2> T2_hasher;
         size_t hashed_seed = T2_hasher(seed);
+        DebugLogger() << "Seed: " << seed << ", Hash: " << hashed_seed;
         return hashed_seed % static_cast<int>(enum_vals_count);
     }
 }
```

For a single player game with seed 0, 200 systems, AI aggression Maniacal, 6 AI, starting species human and everything else set to random, that gave me the following output on OSX:

```
2016-02-19 19:00:16.107874 [debug] Server : Seed: 0shape, Hash: 996847027
2016-02-19 19:00:16.107912 [debug] Server : Seed: 0age, Hash: 1230498160
2016-02-19 19:00:16.107938 [debug] Server : Seed: 0lanes, Hash: 1604476634
2016-02-19 19:00:16.107960 [debug] Server : Seed: 0planets, Hash: 2582775684
2016-02-19 19:00:16.107981 [debug] Server : Seed: 0specials, Hash: 1867141603
2016-02-19 19:00:16.107999 [debug] Server : Seed: 0monsters, Hash: 45877657
2016-02-19 19:00:16.108018 [debug] Server : Seed: 0natives, Hash: 457424265
```

On Linux (Ubuntu 15.10):

```
2016-02-19 18:59:40.812110 [debug] Server : Seed: 0shape, Hash: 3005399786920258995
2016-02-19 18:59:40.812164 [debug] Server : Seed: 0age, Hash: 706246325299568
2016-02-19 18:59:40.812193 [debug] Server : Seed: 0lanes, Hash: 3005401600809393882
2016-02-19 18:59:40.812216 [debug] Server : Seed: 0planets, Hash: 16935225361869965188
2016-02-19 18:59:40.812240 [debug] Server : Seed: 0specials, Hash: 1531096636030865891
2016-02-19 18:59:40.812281 [debug] Server : Seed: 0monsters, Hash: 1530670376820672921
2016-02-19 18:59:40.812304 [debug] Server : Seed: 0natives, Hash: 16935254501819137417
```

What the Niflheim?!?!? Aren't hash functions supposed to yield platform independent results???

I've to call it a day, so I don't have the time to run the same test on Windows. Can you do that? Would be interesting to see if we get yet another different set of hashes.

Anyway, what now? Does boost offer other hash functions we can try, or do we now have to write our own hash algorithms?

**Comment by geoffthemedio** [comment]
I suppose this has come up before...

https://github.com/freeorion/freeorion/issues/294

We don't really need a good hash function for this, though... something small and hand-coded should be fine.

**Comment by geoffthemedio** [comment]
Using the latest...

```
2016-02-19 21:23:10.555438 [debug] Server : Python Universe Generator
2016-02-19 21:23:10.555438 [debug] Server : hashing seed: 0age
2016-02-19 21:23:10.555438 [debug] Server : final hash value: 120 and returning: 0 from 0 to 2
2016-02-19 21:23:10.555438 [debug] Server : hashing seed: 0lanes
2016-02-19 21:23:10.555438 [debug] Server : final hash value: 16 and returning: 1 from 0 to 2
2016-02-19 21:23:10.555438 [debug] Server : hashing seed: 0planets
2016-02-19 21:23:10.555438 [debug] Server : final hash value: 172 and returning: 1 from 0 to 2
2016-02-19 21:23:10.555438 [debug] Server : hashing seed: 0specials
2016-02-19 21:23:10.555438 [debug] Server : final hash value: 115 and returning: 3 from 0 to 3
2016-02-19 21:23:10.555438 [debug] Server : hashing seed: 0monsters
2016-02-19 21:23:10.555438 [debug] Server : final hash value: 11 and returning: 3 from 0 to 3
2016-02-19 21:23:10.555438 [debug] Server : hashing seed: 0natives
2016-02-19 21:23:10.555438 [debug] Server : final hash value: 164 and returning: 0 from 0 to 3
```

or

```
2016-02-19 21:24:33.262086 [debug] Server : Python Universe Generator
2016-02-19 21:24:33.262086 [debug] Server : hashing seed: testage
2016-02-19 21:24:33.262086 [debug] Server : final hash value: 72 and returning: 0 from 0 to 2
2016-02-19 21:24:33.262086 [debug] Server : hashing seed: testlanes
2016-02-19 21:24:33.262086 [debug] Server : final hash value: 159 and returning: 0 from 0 to 2
2016-02-19 21:24:33.262086 [debug] Server : hashing seed: testplanets
2016-02-19 21:24:33.262086 [debug] Server : final hash value: 124 and returning: 1 from 0 to 2
2016-02-19 21:24:33.262086 [debug] Server : hashing seed: testspecials
2016-02-19 21:24:33.262086 [debug] Server : final hash value: 67 and returning: 3 from 0 to 3
2016-02-19 21:24:33.262086 [debug] Server : hashing seed: testmonsters
2016-02-19 21:24:33.262086 [debug] Server : final hash value: 154 and returning: 2 from 0 to 3
2016-02-19 21:24:33.262086 [debug] Server : hashing seed: testnatives
2016-02-19 21:24:33.262086 [debug] Server : final hash value: 116 and returning: 0 from 0 to 3
```

**Comment by Vezzra** [comment]
Yep, that works nicely. I'm getting exact the same hashes for these seeds on OSX and Linux as you. Maps are identical too of course.

Finally! :smiley: So, as far as I'm concerned, good to merge.

**Pull request merged**